modules/cli/src/main/scala/scala/cli/commands/Publish.scala

```diff
@@ -270,43 +315,129 @@ object Publish extends ScalaCommand[PublishOptions] {
 
+    assert(docBuilds.isEmpty || docBuilds.length == builds.length)
+
+    val it = builds.iterator.zip {
+      if (docBuilds.isEmpty) Iterator.continually(None)
+      else docBuilds.iterator.map(Some(_))
+    }
+
     val now = Instant.now()
     val fileSet0 = value {
-      builds
+      it
         // TODO Allow to add test JARs to the main build artifacts
-        .filter(_.scope != Scope.Test)
-        .map { build =>
-          buildFileSet(build, workingDir, now, logger)
+        .filter(_._1.scope != Scope.Test)
+        .map {
+          case (build, docBuildOpt) =>
+            buildFileSet(build, docBuildOpt, workingDir, now, logger)
         }
+        .toVector
         .sequence
         .left.map(CompositeBuildException(_))
         .map(_.foldLeft(FileSet.empty)(_ ++ _))
     }
 
     val ec = builds.head.options.finalCache.ec
 
+    val signerOpt = ConfigMonoid.sum(
+      builds.map(_.options.notForBloopOptions.publishOptions.signer)
+    )
+    val signer: Signer = signerOpt match {
+      case Some(PSigner.Gpg) =>
+        val gpgSignatureIdOpt = ConfigMonoid.sum(
+          builds.map(_.options.notForBloopOptions.publishOptions.gpgSignatureId)
+        )
+        gpgSignatureIdOpt match {
+          case Some(gpgSignatureId) =>
+            val gpgOptions =
+              builds.toList.flatMap(_.options.notForBloopOptions.publishOptions.gpgOptions)
+            GpgSigner(
+              GpgSigner.Key.Id(gpgSignatureId),
+              extraOptions = gpgOptions
+            )
+          case None => NopSigner
+        }
+      case Some(PSigner.BouncyCastle) =>
+        val secretKeyOpt = ConfigMonoid.sum(
+          builds.map(_.options.notForBloopOptions.publishOptions.secretKey)
+        )
+        secretKeyOpt match {
+          case Some(secretKey) =>
+            val passwordOpt = ConfigMonoid.sum(
+              builds.map(_.options.notForBloopOptions.publishOptions.secretKeyPassword)
+            )
+            val getLauncher: Supplier[NioPath] = { () =>
+              val archiveCache = builds.headOption
+                .map(_.options.archiveCache)
+                .getOrElse(ArchiveCache())
+              PgpExternalCommand.launcher(archiveCache, None, logger) match {
+                case Left(e)      => throw new Exception(e)
+                case Right(value) => value.wrapped
+              }
+            }
+            (new BouncycastleSignerMaker).get(
+              passwordOpt.orNull,
+              secretKey.wrapped,
+              getLauncher,
+              logger
+            )
+          case None => NopSigner
+        }
+      case None => NopSigner
+    }
+    val signerLogger =
+      new InteractiveSignerLogger(new OutputStreamWriter(System.err), verbosity = 1)
+    val signRes = signer.signatures(
+      fileSet0,
+      now,
+      ChecksumType.all.map(_.extension).toSet,
+      Set("maven-metadata.xml"),
+      signerLogger
+    )
+
+    val fileSet1 = value {
```

Comment:
can we call it more meaningfull then `fileSet1`?

Reply:
`fileSet0`, `fileSet1`, `fileSet2`, and `finalFileSet` each add some more data, and are there to avoid using mutable variables. Maybe we should use a mutable `fileSet` variable here…